codegen/src/main/scala/bsp/codegen/docs/MarkdownRenderer.scala

```diff
@@ -25,69 +32,88 @@ class MarkdownRenderer private (tree: DocTree, visited: MSet[ShapeId]) {
     if (id.getNamespace == "smithy.api") {
       return empty
     }
 
-    tree.docNodes(id) match {
-      case OperationDocNode(operation, inputNode, outputNode) =>
+    val node = tree.structures(id)
+
+    if (visited.contains(id)) empty
+    else {
+      visited.add(id)
+      lines(
+        renderStructureNodeDef(node.definition),
+        node.definition.members
+          .filter(_.getNamespace == id.getNamespace)
+          .foldMap(renderStructureNode)
+      )
+    }
+  }
+
+  def renderStructureNodeDef(definition: Def): Lines = {
+    lines(
+      s"#### ${definition.shapeId.getName()}",
+      newline,
+      documentation(definition.hints),
+      newline,
+      tsBlock(definition),
+      newline
+    )
+  }
+
+  def renderOperationNode(node: OperationDocNode): Lines = {
+    node match {
+      case OperationDocNode(operation, input, output) =>
         lines(
           s"### ${operation.name}: ${methodTpe(operation)}",
           newline,
           documentation(operation.hints),
           newline,
           s"- method: `${operation.jsonRPCMethod}`",
-          inputNode.foldMap(n => s"- params: `${n.getName()}`"),
-          outputNode.foldMap(n => s"- result: `${n.getName()}`"),
-          inputNode.foldMap(renderNode),
-          outputNode.foldMap(renderNode),
-          newline
+          input.foldMap(n => s"- params: `${n.getName()}`"),
+          output.foldMap(n => s"- result: `${n.getName()}`"),
+          newline,
+          input.foldMap(renderStructureNode),
+          output.foldMap(renderStructureNode)
         )
+    }
+
+  }
+
+  def renderServiceNode(node: ServiceDocNode): Lines = {
+    node match {
       case ServiceDocNode(shapeId, operations) =>
-        if (shapeId.getName().toLowerCase().contains("server")) {
-          lines(
-            s"## BSP Server remote interface",
-            newline,
-            operations.foldMap(renderNode)
-          )
+        val header = if (shapeId.getName().toLowerCase().contains("server")) {
+          s"## BSP Server remote interface"
         } else {
-          lines(
-            s"## BSP Client remote interface",
-            newline,
-            operations.foldMap(renderNode)
-          )
+          s"## BSP Client remote interface"
         }
-      case ShapeDocNode(definition, members) =>
-        if (visited.contains(definition.shapeId)) empty
-        else
-          {
-            visited.add(definition.shapeId)
-            lines(
-              s"#### ${definition.shapeId.getName()}",
-              newline,
-              documentation(definition.hints),
-              newline,
-              tsBlock(definition),
-              newline
-            )
-          } ++ lines(
-            members.foldMap(renderNode)
-          )
+        lines(
+          header,
+          newline,
+          operations
+            .map(tree.operations)
+            .foldMap(renderOperationNode)
```

Comment:
Feel free to ignore this since you didn't add it here, but I'm not sure how others feel about using deps transitively without explicitly declaring them in the build. I see we're using `foldMap` here from cats, but we never define a cats dependency. Again, no strong feelings on this but I typically like to declare something if we're explicitly using it.